File: 100/164 - Maximum Gap.java
public class Solution {
    public int maximumGap(int[] nums) {
        int minNum, maxNum, len, i, p, ret;
        int[][] buckets;
        
        if (nums == null || nums.length < 2)
            return 0;
        minNum = nums[0];
        maxNum = nums[0];
         
        for (int num : nums) {
            minNum = Math.min(minNum, num);
            maxNum = Math.max(maxNum, num);
        }
        
        len = (maxNum - minNum) / (nums.length - 1) + 1;
        buckets = new int[(maxNum - minNum) / len + 1][];
        
        for (int num : nums) {
            i = (num - minNum) / len;
            if (buckets[i] == null) {
                buckets[i] = new int[2];
                buckets[i][0] = num;
                buckets[i][1] = num;
            } else {
                if (num < buckets[i][0])
                    buckets[i][0] = num;
                else if (num > buckets[i][1])
                    buckets[i][1] = num;
            }
        }
        
        ret = buckets[0][1] - buckets[0][0];
        p = 0;
        for (i = 1; i < buckets.length; ++i) {
            if (buckets[i] != null) {
                ret = Math.max(ret, buckets[i][0] - buckets[p][1]);
                p = i;
            }
        }
        
        return ret;
    }
}

File: csharp/164.maximum-gap.cs
/*
 * @lc app=leetcode id=164 lang=csharp
 *
 * [164] Maximum Gap
 */
public class Solution {
    public int MaximumGap(int[] nums) {
        int n = nums.Length;
        if (n < 2) {
            return 0;
        }
        int min = int.MaxValue, max = int.MinValue;
        foreach (int num in nums) {
            min = Math.Min(min, num);
            max = Math.Max(max, num);
        }
        int size = (max - min) / n + 1, count = (max - min) / size + 1;
        int[] mins = new int[count], maxs = new int[count];
        for (int i = 0; i < count; ++i) {
            mins[i] = int.MaxValue;
            maxs[i] = int.MinValue;
        }
        foreach (int num in nums) {
            int i = (num - min) / size;
            mins[i] = Math.Min(mins[i], num);
            maxs[i] = Math.Max(maxs[i], num);
        }
        int ret = 0, prev = 0;
        for (int i = 1; i < count; ++i) {
            if (mins[i] == int.MaxValue) {
                continue;
            }
            ret = Math.Max(ret, mins[i] - maxs[prev]);
            prev = i;
        }
        return ret;
    }
}

